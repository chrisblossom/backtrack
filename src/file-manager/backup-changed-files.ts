import path from 'path';
import { rootPath } from '../config/paths';
import { backupFile } from '../utils/backup-file';
import log from '../utils/log';

import { ParsedFiles, FileStats } from '../types';

async function backupChangedFiles(
	files: ParsedFiles,
	previousStats: FileStats = {},
) {
	const backupNewFiles = files.dest.files
		.map((destFile) => {
			/**
			 * This only handles new files
			 */
			const previousFileHash = previousStats[destFile];
			if (previousFileHash) {
				return null;
			}

			const destFileIndex = files.dest.files.indexOf(destFile);
			const srcFile = files.src.files[destFileIndex];
			const srcHash = files.src.hash[srcFile];
			const destHash = files.dest.hash[destFile];
			const allowChanges = files.dest.allowChanges[destFile];
			const destExists = !!destHash;

			/**
			 * Don't backup if file doesn't exist or the files are the same
			 */
			if (!destExists || srcHash === destHash) {
				return null;
			}

			/**
			 * Do not backup file if changes are allowed
			 */
			if (allowChanges === true) {
				return null;
			}

			const originalFilePath = files.dest.absolute[destFile];
			return backupFile(originalFilePath).then((backupResult) => {
				if (backupResult) {
					const relativeBackupFilePath = path.relative(
						rootPath,
						backupResult.file,
					);

					const relativeFilePath = path.relative(
						rootPath,
						originalFilePath,
					);

					log.warn(
						`${relativeFilePath} already exists. Moving to ${relativeBackupFilePath}`,
					);
				}

				return backupResult;
			});
		})
		.filter(Boolean);

	const backupOldFiles = Object.keys(previousStats)
		.map((destFile) => {
			const destHash = files.dest.hash[destFile];

			const previousFileHash = previousStats[destFile];
			const destExists = !!destHash;

			const destFileIndex = files.dest.files.indexOf(destFile);

			/**
			 * Need to check for undefined because previousStats source file could have been removed
			 */
			const srcFile =
				destFileIndex !== -1
					? files.src.files[destFileIndex]
					: undefined;

			const srcHash = srcFile ? files.src.hash[srcFile] : undefined;

			/**
			 * Do not backup file if changes are allowed
			 */
			const allowChanges = files.dest.allowChanges[destFile];
			if (allowChanges === true) {
				return null;
			}

			/**
			 * Don't backup if dest hasn't changed
			 */
			if (
				destExists === false ||
				previousFileHash === destHash ||
				// already correct hash
				srcHash === destHash
			) {
				return null;
			}

			/**
			 * Move current destFile if modified
			 */
			const originalFilePath = files.dest.absolute[destFile];

			return backupFile(originalFilePath).then((backupResult) => {
				if (backupResult) {
					const relativeBackupFilePath = path.relative(
						rootPath,
						backupResult.file,
					);

					const relativeFilePath = path.relative(
						rootPath,
						originalFilePath,
					);

					log.warn(
						`${relativeFilePath} hash mismatch. Moving to ${relativeBackupFilePath}`,
					);
				}

				return backupResult;
			});
		})
		.filter(Boolean);

	// @ts-ignore
	await Promise.all([...backupNewFiles, ...backupOldFiles]);
}

export { backupChangedFiles };
