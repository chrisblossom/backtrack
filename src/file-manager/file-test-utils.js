/* @flow */

/* eslint-disable arrow-body-style */

import type { FileManager } from '../types.js';

const filesPreprocessor = (args) =>
    // $FlowIgnore
    require('../options-file/files-preprocessor').filesPreprocessor(args);
const filesParser = (args) =>
    // $FlowIgnore
    require('../options-file/files-post-processor').filesPostProcessor(args);
const getFileStats = (parsedFiles) =>
    require('./get-file-stats').getFileStats(parsedFiles);

function fileStats(previousFiles: FileManager) {
    const normalize = filesPreprocessor({ value: previousFiles });

    const parsedPreviousFiles = filesParser({ value: normalize });

    const previousStats = getFileStats(parsedPreviousFiles).files;

    return previousStats;
}

function fileInfo(files: FileManager = [], previousFiles: FileManager = []) {
    const previousStats = fileStats(previousFiles);
    const normalize = filesPreprocessor({ value: files });

    const parsedFiles = filesParser({ value: normalize });

    return {
        previousStats,
        parsedFiles,
    };
}

export { fileInfo, fileStats };
