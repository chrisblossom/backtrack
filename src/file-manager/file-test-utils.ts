/* eslint-disable arrow-body-style,@typescript-eslint/no-unsafe-member-access,@typescript-eslint/no-unsafe-return,@typescript-eslint/no-unsafe-call,@typescript-eslint/no-unsafe-assignment */

import { FileManager, FileManagerStats } from '../types';

const filesPreprocessor = (args: any) =>
	require('../options-file/files-preprocessor').filesPreprocessor(args);
const filesParser = (args: any) =>
	require('../options-file/files-post-processor').filesPostProcessor(args);
const getFileStats = (parsedFiles: any) =>
	require('./get-file-stats').getFileStats(parsedFiles);

function fileStats(previousFiles: FileManager): FileManagerStats {
	const normalize = filesPreprocessor({ value: previousFiles });

	const parsedPreviousFiles = filesParser({ value: normalize });

	const previousStats = getFileStats(parsedPreviousFiles).files;

	return previousStats;
}

interface FileInfo {
	previousStats: FileManagerStats;
	parsedFiles: FileManager;
}

function fileInfo(
	files: FileManager = [],
	previousFiles: FileManager = [],
): FileInfo {
	const previousStats = fileStats(previousFiles);
	const normalize = filesPreprocessor({ value: files });

	const parsedFiles = filesParser({ value: normalize });

	return {
		previousStats,
		parsedFiles,
	};
}

export { fileInfo, fileStats };
