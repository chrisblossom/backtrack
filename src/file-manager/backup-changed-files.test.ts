import path from 'path';
import { fileInfo } from './file-test-utils';

const backupChangedFiles = (files: any, previousStats: any) =>
    require('./backup-changed-files').backupChangedFiles(files, previousStats);

describe('backupChangedFiles', () => {
    const cwd = process.cwd();
    let move: any;

    beforeEach(() => {
        move = require.requireMock('fs-extra').move;

        jest.mock('fs-extra', () => ({
            move: jest.fn(() => Promise.resolve()),
        }));

        jest.mock('../utils/log', () => ({
            warn: jest.fn(),
        }));
    });

    afterEach(() => {
        process.chdir(cwd);
    });

    test('backups files with different hashes', async () => {
        const dir = path.resolve(__dirname, '__sandbox__/stats1/');
        process.chdir(dir);

        const files = [
            {
                src: path.resolve(dir, 'file1.js'),
                dest: 'nested/other.js',
            },
        ];

        const { parsedFiles, previousStats } = fileInfo(files);

        await backupChangedFiles(parsedFiles, previousStats);

        expect(move.mock.calls).toMatchSnapshot();
    });

    test('does not backup if changes allowed', async () => {
        const dir = path.resolve(__dirname, '__sandbox__/stats1/');
        process.chdir(dir);

        const previousFiles = [
            {
                src: path.resolve(dir, 'nested/other.js'),
                dest: 'nested/other.js',
                allowChanges: true,
            },
        ];

        const files = [
            {
                src: path.resolve(dir, 'file1.js'),
                dest: 'nested/other.js',
                allowChanges: true,
            },
        ];

        const { parsedFiles, previousStats } = fileInfo(files, previousFiles);

        await backupChangedFiles(parsedFiles, previousStats);

        expect(move.mock.calls).toEqual([]);
    });

    test('does not backup if changes allowed and not previously managed', async () => {
        const dir = path.resolve(__dirname, '__sandbox__/stats1/');
        process.chdir(dir);

        const files = [
            {
                src: path.resolve(dir, 'file1.js'),
                dest: 'nested/other.js',
                allowChanges: true,
            },
        ];

        const { parsedFiles, previousStats } = fileInfo(files);

        await backupChangedFiles(parsedFiles, previousStats);

        expect(move.mock.calls).toEqual([]);
    });

    test('no previousStats - backup if file exists', async () => {
        const dir = path.resolve(__dirname, '__sandbox__/stats1/');
        process.chdir(dir);

        const files = [
            {
                src: path.resolve(dir, 'file1.js'),
                dest: 'nested/other.js',
            },
        ];

        const { parsedFiles, previousStats } = fileInfo(files);

        await backupChangedFiles(parsedFiles, previousStats);

        expect(move.mock.calls).toMatchSnapshot();
    });

    test('no previousStats - dest does not exist', async () => {
        const dir = path.resolve(__dirname, '__sandbox__/stats1/');
        process.chdir(dir);

        const files = [
            {
                src: path.resolve(dir, 'file1.js'),
                dest: 'nested/missing.js',
                allowChanges: true,
            },
        ];

        const { parsedFiles, previousStats } = fileInfo(files);

        await backupChangedFiles(parsedFiles, previousStats);

        expect(move.mock.calls).toEqual([]);
    });

    test('no previousStats - dest is same file', async () => {
        const dir = path.resolve(__dirname, '__sandbox__/stats1/');
        process.chdir(dir);

        const files = [
            {
                src: path.resolve(dir, 'file1.js'),
                dest: 'file1.js',
                allowChanges: true,
            },
        ];

        const { parsedFiles, previousStats } = fileInfo(files);

        await backupChangedFiles(parsedFiles, previousStats);

        expect(move.mock.calls).toEqual([]);
    });

    test('does nothing if updated file matches target file hash', async () => {
        const dir = path.resolve(__dirname, '__sandbox__/stats1/');
        process.chdir(dir);

        const files = [
            {
                src: path.resolve(dir, 'nested/file1.js'),
                dest: 'file1.js',
            },
        ];

        const { parsedFiles } = fileInfo(files);

        const previousStats = {
            'file1.js': 'fake_hash',
        };

        const result = await backupChangedFiles(parsedFiles, previousStats);

        expect(move.mock.calls).toEqual([]);
        expect(result).toBeUndefined();
    });

    test('does nothing if file does not exist', async () => {
        const dir = path.resolve(__dirname, '__sandbox__/stats1/');
        process.chdir(dir);

        const files = [
            {
                src: path.resolve(dir, 'nested/inside.js'),
                dest: 'nested/does_not_exist.js',
            },
        ];

        const { parsedFiles, previousStats } = fileInfo(files);

        const result = await backupChangedFiles(parsedFiles, previousStats);

        expect(move.mock.calls).toEqual([]);
        expect(result).toBeUndefined();
    });

    test('does nothing if no files changed', async () => {
        const dir = path.resolve(__dirname, '__sandbox__/stats1/');
        process.chdir(dir);

        const files = [
            {
                src: path.resolve(dir, 'nested/other.js'),
                dest: 'nested/other.js',
            },
        ];

        const { parsedFiles, previousStats } = fileInfo(files);

        const result = await backupChangedFiles(parsedFiles, previousStats);

        expect(move.mock.calls).toEqual([]);
        expect(result).toBeUndefined();
    });
});
