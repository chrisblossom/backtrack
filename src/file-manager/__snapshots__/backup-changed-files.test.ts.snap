// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`backupChangedFiles backups files with different hashes 1`] = `
[
  [
    "<PROJECT_ROOT>/nested/other.js",
    "<PROJECT_ROOT>/nested/other.js-backup-44c09b27.js",
    {
      "overwrite": false,
    },
  ],
]
`;

exports[`backupChangedFiles no previousStats - backup if file exists 1`] = `
[
  [
    "<PROJECT_ROOT>/nested/other.js",
    "<PROJECT_ROOT>/nested/other.js-backup-44c09b27.js",
    {
      "overwrite": false,
    },
  ],
]
`;
